Docs 4427 dh

dhirsch1 opened the pull request:
Rewrote the Oauth Authorization Grant Types topic and added the topic to the Mule 4 TOC in the Security section. I also moved the TLS Configuration topic to that section.

dhirsch1 committed:
Added the v3.9 file to v4.1
Added the v3.9 file to v4.1

dhirsch1 committed:
Some more wording updates
Some more wording updates

dhirsch1 committed:
Polishing the wording
Polishing the wording

dhirsch1 committed:
Commit in case commit through git tab in Atom did not work.
Commit in case commit through git tab in Atom did not work.

dhirsch1 committed:
Fixed list formatting in athorization-grant-types, added topic to TOC, moved TLS to the Security section  TOC
Fixed list formatting in athorization-grant-types, added topic to TOC, moved TLS to the Security section  TOC

lucianoRM commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,101 @@
+endif::[]
+:keywords: http, authentication, security, users, connectors, anypoint, studio, oauth
+
+There are four types of authorization grants that an OAuth consumer (that is, a client app) can use to obtain access to a protected resource from an OAuth service provider. The types of authorization grants are listed below according to their security level, in descending order, with the Authorization Code grant type being the most secure.
```
I don't know if one grant type is more secure than the other, they are just used in different scenarios.

aiannucci commented:
@lucianoRM can you please take a look?

lucianoRM commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,101 @@
+
+The Authorization Code grant type uses an authorization server (responsible for confirming and granting permission to access the protected resource) and a resource server (responsible for providing access to the protected resource). Though described as independent servers, the authorization and resource servers reside on the same Mule server. With the Authorization Code grant type, a client app m ...
+
+The Authorization Code grant type is especially secure, because it authenticates the client and transmits the access token without exposing the it to unauthorized parties, including the resource owner.
```
exposing the it -> exposing it

lucianoRM commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,101 @@
+
+== Authorization Code
+
+The Authorization Code grant type uses an authorization server (responsible for confirming and granting permission to access the protected resource) and a resource server (responsible for providing access to the protected resource). Though described as independent servers, the authorization and resource servers reside on the same Mule server. With the Authorization Code grant type, a client app m ...
```
a client app must ask for an authorization code from the service provider before it can ask for access to the protected resources. -> 
a client app must ask for an authorization code from the service provider before it can ask for the token that will grant him access to the protected resources.

lucianoRM commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,101 @@
+
+. Client app calls the service provider's authorization server, providing its client ID to confirm that it is a registered client of the service provider.
+
+. Client app uses a login page supplied by the service provider to ask that the resource owner authenticate itself to the service provider.
```
TO authenticate itself

dhirsch1 replied:
Are you trying to say this? Client app uses a login page supplied by the service provider to authenticate itself to the service provider.

lucianoRM replied:
sorry i thought it was a typo and the word TO was missing. Shouldn't it be `that the resource owner authenticateS itself`?

dhirsch1 replied:
No, the wording is correct (as long as it's the resource owner that needs to authenticate itself).

lucianoRM commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,101 @@
+= OAuth Authorization Grant Types
```
Reference OAuth2 RFC:
https://tools.ietf.org/html/rfc6749

dhirsch1 committed:
Added edits from Luciano.
Added edits from Luciano.

dhirsch1 committed:
Unhid the reference to the RFC.
Unhid the reference to the RFC.

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+There are four types of authorization grants that an OAuth consumer (that is, a client app) can use to obtain access to a protected resource from an OAuth service provider:
+
+* <<Authorization Code>>
```
Recommend using anchors for the section links instead of the section title because anchors are less fragile. I'll Slack you about it.

dhirsch1 replied:
OK

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+* <<Client Credentials>>
+
+For information about the OAuth2 RFC, see https://tools.ietf.org/html/rfc6749.
```
We avoid links to 3rd party sites because we don't control them. You could perhaps reference the searchable name, "RFC 6749" and title, "The OAuth 2.0 Authorization Framework."

dhirsch1 replied:
OK

dhirsch1 replied:
OK

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+For information about the OAuth2 RFC, see https://tools.ietf.org/html/rfc6749.
+
+////
```
Is there a JIRA for the link to the Mule Secure Token Service OAuth 2.0 Provider page? Please add one.

dhirsch1 replied:
I will create one if I didn't do that already.

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+== Authorization Code
+
+The Authorization Code grant type uses an authorization server (responsible for confirming and granting permission to access the protected resource) and a resource server (responsible for providing access to the protected resource). Though described as independent servers, the authorization and resource servers reside on the same Mule server. With the Authorization Code grant type, a client app m ...
```
"ask for" -> "request"  (typically, programmatic "asks" are referred to as "requests")

"can ask for" -> "can request"

dhirsch1 replied:
OK

dhirsch1 replied:
OK

dhirsch1 replied:
OK

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+The Authorization Code grant type uses an authorization server (responsible for confirming and granting permission to access the protected resource) and a resource server (responsible for providing access to the protected resource). Though described as independent servers, the authorization and resource servers reside on the same Mule server. With the Authorization Code grant type, a client app m ...
+
+The Authorization Code grant type is especially secure, because it authenticates the client and transmits the access token without exposing it to unauthorized parties, including the resource owner.
```
"secure, because" -> "secure because" See, for ex, https://www.grammarly.com/blog/comma-before-because/

dhirsch1 replied:
Given that this is a very long sentence, I think leaving the comma in helps with clarity.

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+An OAuth dance that uses an Authorization Code grant type works as follows:
+
+. Client app wants to use a protected resource housed by the service provider, so the client initiates a request to access the resource.
```
Suggest word changes:
"wants to" is a bit anthropomorphic however tempting it is to use that terminology. Recommend avoiding that when feasible. 

Is "housed" is a bit unusual in this context? Maybe "stored" or "held" or "controlled" or some such?

dhirsch1 replied:
Left "wants" but changed "housed" to "controlled."

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+. Client app uses a login page supplied by the service provider to authenticate itself to the service provider.
+
+. After confirming that the client ID is valid (that is, the client app had previously registered with the service provider), and authenticating the resource owner's login credentials, the service provider returns an authorization code.
```
"provider), and" -> "provider) and"

dhirsch1 replied:
Given that this is a very long sentence, I think leaving the comma in helps with clarity.

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+. After confirming that the client ID is valid (that is, the client app had previously registered with the service provider), and authenticating the resource owner's login credentials, the service provider returns an authorization code.
+
+. Client app calls the authorization server again, and provides its authorization code, client ID (again), and client secret.
```
"again, and" -> "again and"

dhirsch1 replied:
Given that this is a very long sentence, I think leaving the comma in helps with clarity.

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+. Service provider returns an access token.
+
+. Client app calls the resource server, providing the access token, to request the protected resource.
```
"token, to" -> "token to"

dhirsch1 replied:
OK

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+. Client app calls the service provider's authorization server, providing its client ID to confirm that it is a registered client of the service provider.
+
+. Client app uses a login page supplied by the service provider to ask that the resource owner authenticate itself to the service provider.
```
"to ask that the ..." -> "so that the resource owner can provide authentication credentials to the service provider" or some such?

dhirsch1 replied:
Changed to "to request that..."

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+. Client app uses a login page supplied by the service provider to ask that the resource owner authenticate itself to the service provider.
+
+. After confirming that the client ID is valid (that is, the client app had previously registered with the service provider), and authenticating the resource owner's login credentials, the service provider returns an access token.
```
"provider), and" -> "provider) and"

dhirsch1 replied:
Given that this is a very long sentence, I think leaving the comma in helps with clarity.

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+. After confirming that the client ID is valid (that is, the client app had previously registered with the service provider), and authenticating the resource owner's login credentials, the service provider returns an access token.
+
+. Client app calls the resource server, providing the access token, to request the protected resource.
```
"token, to" -> "token to"

dhirsch1 replied:
OK

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+== Resource Owner Password Credentials
+
+With the Resource Owner Password Credentials grant type, a client application demands that the resource owner share their service provider login credentials. The client app then uses these credentials to access the service provider and log into the resource owner's account. The service provider accepts the valid login credentials and grants the client app full access to the resource owner's accou ...
```
"share their" -> "share its"

dhirsch1 replied:
OK

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+This type of authorization optimizes performance at the expense of sharing secure access credentials. The resource owner must trust the client app not to share access to private data.
+
+An OAuth dance uses the Resource Owner Password Credentials grant type works as follows:
```
"works as follows" -> "as follows" 

or maybe "An OAuth dance uses" -> "An OAuth dance that uses"

dhirsch1 replied:
OK

dhirsch1 replied:
OK

dhirsch1 replied:
OK

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+. Resource owner provides access credentials to the client app.
+
+. Client app calls the service provider to request an access token. During the call, the client provides the client ID to confirm that it is a registered client of the service provider, and also provides the access credentials for the protected resource.
```
"and also" -> "and it"

dhirsch1 replied:
Changed to two sentences.

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+. Service provider authenticates the client app and validates the access credentials. Then it issues an access token.
+
+. Client app calls the resource server, providing the access token, to request access to the protected resource.
```
"token, to" -> "token to"

dhirsch1 replied:
OK

dhirsch1 replied:
OK

dhirsch1 replied:
OK

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+== Client Credentials
+
+With the Client Credentials grant type, a client app requests access to data that is not associated with a particular resource owner. Because no protected resources are involved, a service provider must only verify that the client as valid. This type of OAuth interaction is sometimes referred to as "two-legged OAuth" because it involves only two roles: the service provider and a client app.  The  ...
```
"no protected resources" -> maybe "no password-protected resources" or something like that? Looks like it's a case where authentication is not required, only authorization. Is that correct?

dhirsch1 replied:
OK

dukesphere commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+. After confirming that the client ID is valid (that is, the client app had previously registered with the service provider), the service provider returns an access token.
+
+. Client app calls the resource server, providing the access token, to request the protected resource.
```
"token, to" -> "token to"?

dhirsch1 replied:
OK

dhirsch1 replied:
OK

dhirsch1 replied:
OK

dukesphere reviewed (changes requested):
comma quibbles, some assessments of terminology, and some style recommendations

lucianoRM commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+== Client Credentials
+
+With the Client Credentials grant type, a client app requests access to data that is not associated with a particular resource owner. Because no protected resources are involved, a service provider must only verify that the client as valid. This type of OAuth interaction is sometimes referred to as "two-legged OAuth" because it involves only two roles: the service provider and a client app.  The  ...
```
I don't know if this is correct. 
From the RFC: 
`The client credentials (or other forms of client authentication) can
   be used as an authorization grant when the authorization scope is
   limited to the protected resources under the control of the client,
   or to protected resources previously arranged with the authorization
   server.`

There are protected resources involved

dhirsch1 committed:
Updated with reviewer comments
Updated with reiewer comments

lucianoRM commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+* <<Client Credentials>>
+
+For information about the OAuth2 RFC, see https://tools.ietf.org/html/rfc6749.
```
It should be: "For information about the OAuth2 protocol, see the RFC"

dhirsch1 replied:
In the most recent commit I have "For information about the OAuth authorization framework, see RFC 6749, "The OAuth 2.0 Authorization Framework." 

Is that OK?

lucianoRM replied:
yes

lucianoRM commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+== Authorization Code
+
+The Authorization Code grant type uses an authorization server (responsible for confirming and granting permission to access the protected resource) and a resource server (responsible for providing access to the protected resource). Though described as independent servers, the authorization and resource servers reside on the same Mule server. With the Authorization Code grant type, a client app m ...
```
the protected resource -> protected resources

dhirsch1 replied:
Will change.

lucianoRM commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+. After confirming that the client ID is valid (that is, the client app had previously registered with the service provider), and authenticating the resource owner's login credentials, the service provider returns an authorization code.
+
+. Client app calls the authorization server again, and provides its authorization code, client ID (again), and client secret.
```
its authorization code -> the authorization code

dhirsch1 replied:
OK

lucianoRM commented on modules/ROOT/pages/authorization-grant-types.adoc:
```diff
@@ -0,0 +1,103 @@
+
+This type of authorization optimizes performance at the expense of sharing secure access credentials. The resource owner must trust the client app not to share access to private data.
+
+An OAuth dance uses the Resource Owner Password Credentials grant type works as follows:
```
an OAuth dance uses -> an OAuth dance that uses

dhirsch1 replied:
This was already corrected in the updated commit,

dhirsch1 committed:
Added one more edit: changed
Added one more edit: changed "its token" to "the token" "its token" to "the token"
